From 84a263a0f7342ca254b40d20d27bea078ce80eec Mon Sep 17 00:00:00 2001
From: Juan Lozano <lozanotux@gmail.com>
Date: Sun, 2 Aug 2020 21:10:30 -0300
Subject: [PATCH 01/17] Fix icon alignment and space between them, and improove
 background color

---
 src/Application.vala | 27 +++++++++++----------------
 1 file changed, 11 insertions(+), 16 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index 49bddb3..f6de636 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -96,13 +96,8 @@ public class LightPadWindow : Widgets.CompositedWindow {
         // Add top bar
         var bottom = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
 
-        this.top_spacer = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 20);
-        this.top_spacer.realize.connect ( () => { this.top_spacer.visible = true; } );
-        this.top_spacer.can_focus = true;
-        bottom.pack_start (this.top_spacer, false, false, 0);
-
         // Searchbar
-        this.searchbar = new LightPad.Frontend.Searchbar ("...");
+        this.searchbar = new LightPad.Frontend.Searchbar ("Search");
         message ("Searchbar created!");
         this.searchbar.changed.connect (this.search);
 
@@ -114,8 +109,8 @@ public class LightPadWindow : Widgets.CompositedWindow {
         container.pack_start (bottom, false, true, 32);
 
         this.grid = new Gtk.Grid();
-        this.grid.set_row_spacing (70);
-        this.grid.set_column_spacing (30);
+        this.grid.set_row_spacing (30);
+        this.grid.set_column_spacing (0);
         this.grid.set_halign (Gtk.Align.CENTER);
 
         // Make icon grid and populate
@@ -124,14 +119,17 @@ public class LightPadWindow : Widgets.CompositedWindow {
             this.grid_x = 5;
             this.grid_y = 5;
         } else if (monitor_dimensions.height == 600) { // Netbook 1024x600px
-            this.grid_y = 5;
-            this.grid_x = 3;
-        } else if (monitor_dimensions.height == 1080) { // Full HD 1920x1080px
+            this.grid_y = 6;
+            this.grid_x = 4;
+        } else if (monitor_dimensions.height == 720) { // HD 1280x720px
             this.grid_y = 7;
             this.grid_x = 5;
+        } else if (monitor_dimensions.height == 1080) { // Full HD 1920x1080px
+            this.grid_y = 9;
+            this.grid_x = 7;
         } else { // Monitor 16:9
             this.grid_y = 6;
-            this.grid_x = 4;
+            this.grid_x = 5;
         }
 
         // Initialize the grid
@@ -221,7 +219,6 @@ public class LightPadWindow : Widgets.CompositedWindow {
 
                 item.button_press_event.connect ( () => { item.grab_focus (); return true; } );
                 item.enter_notify_event.connect ( () => { item.grab_focus (); return true; } );
-                item.leave_notify_event.connect ( () => { this.top_spacer.grab_focus (); return true; } );
                 item.button_release_event.connect ( () => {
                     try {
                         int child_index = this.children.index (item);
@@ -343,9 +340,7 @@ public class LightPadWindow : Widgets.CompositedWindow {
             widget.get_allocation (out size);
 
             var linear_gradient = new Cairo.Pattern.linear (size.x, size.y, size.x, size.y + size.height);
-            linear_gradient.add_color_stop_rgba (0.0, 0.0, 0.0, 0.0, 1);
-            linear_gradient.add_color_stop_rgba (0.50, 0.0, 0.0, 0.0, 0.90);
-            linear_gradient.add_color_stop_rgba (0.99, 0.0, 0.0, 0.0, 0.80);
+            linear_gradient.add_color_stop_rgba (0.0, 0.30, 0.30, 0.30, 1);
 
             context.set_source (linear_gradient);
             context.paint ();

From 1fedf75deead7f6c8159683aada90d4ddb5538fd Mon Sep 17 00:00:00 2001
From: m <m@metaquanta.com>
Date: Tue, 1 Dec 2020 06:08:17 -0500
Subject: [PATCH 02/17] More robust monitor selection and geometry

---
 src/Application.vala | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index f6de636..2c043c8 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -52,8 +52,12 @@ public class LightPadWindow : Widgets.CompositedWindow {
     public Gdk.Pixbuf image_pf;
 
     public LightPadWindow () {
-        Gdk.Screen default_screen = Gdk.Screen.get_default ();
-        monitor_dimensions = default_screen.get_display ().get_primary_monitor ().get_geometry ();
+        // There isn't always a primary monitor.
+        Gdk.Monitor monitor = get_display ().get_primary_monitor () ?? get_display ().get_monitor (0);
+        Gdk.Rectangle pixel_geo = monitor.get_geometry ();
+        // get_geometry() returns "device pixels", but we need "application pixels".
+        monitor_dimensions.width = pixel_geo.width / monitor.get_scale_factor ();
+        monitor_dimensions.height = pixel_geo.height / monitor.get_scale_factor ();
 
         // Window properties
         this.set_title ("LightPad");

From 9416cdbad75f4a0e6f1bb17cb8f54663c7db97c2 Mon Sep 17 00:00:00 2001
From: DEB Libre <libredeb@gmail.com>
Date: Thu, 28 Jan 2021 00:02:04 -0300
Subject: [PATCH 03/17] Update debian/control

Fix misspelled word in "Description" tag to "Written"
---
 debian/control | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/debian/control b/debian/control
index 781c701..26d6a14 100644
--- a/debian/control
+++ b/debian/control
@@ -25,4 +25,4 @@ Depends: ${misc:Depends}, ${shlibs:Depends},
          gnome-menus (>= 3.13.0),
          libwnck-3-0 (>= 3.20.0),
          xterm
-Description: LightPad is a lightweight, simple and powerful application launcher. Writen in GTK+ 3.0. It is also Wayland compatible.
+Description: LightPad is a lightweight, simple and powerful application launcher. Written in GTK+ 3.0. It is also Wayland compatible.

From 5591802094743377f04c645563f8dd6ad108aa68 Mon Sep 17 00:00:00 2001
From: DEB Libre <libredeb@gmail.com>
Date: Thu, 28 Jan 2021 00:03:22 -0300
Subject: [PATCH 04/17] Update rpm/spec file

Fix misspelled word in %description section to "Written"
---
 rpm-build.spec | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/rpm-build.spec b/rpm-build.spec
index 852e798..fa905ee 100644
--- a/rpm-build.spec
+++ b/rpm-build.spec
@@ -35,7 +35,7 @@ Requires:   	xterm
 
 %description
 LightPad is a lightweight, simple and powerful application launcher.
-Writen in GTK+ 3.0. It is also Wayland compatible.
+Written in GTK+ 3.0. It is also Wayland compatible.
 
 %prep
 %setup -q

From 78bd08c2ddbd630976b7421e283dff8cf9f2e908 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ph=C3=BAc=20La?= <phucla@outlook.com>
Date: Wed, 19 May 2021 00:31:22 +0700
Subject: [PATCH 05/17] add basic config

---
 .gitignore               |   1 +
 post_install.py          |   6 ++-
 src/Application.vala     |   7 ++-
 src/Config.vala          | 109 +++++++++++++++++++++++++++++++++++++++
 src/Resources.vala       |   1 +
 src/Widgets/AppItem.vala |   5 +-
 src/config               |  15 ++++++
 7 files changed, 141 insertions(+), 3 deletions(-)
 create mode 100644 src/Config.vala
 create mode 100644 src/config

diff --git a/.gitignore b/.gitignore
index cacb413..da88a31 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 *~
 build/
 *lightpad-*.pkg.*
+how_to_run.txt
diff --git a/post_install.py b/post_install.py
index 29d088a..5dc4860 100644
--- a/post_install.py
+++ b/post_install.py
@@ -7,4 +7,8 @@
 
 if not os.environ.get('DESTDIR'):
     print('Updating icon cache...')
-    subprocess.call(['gtk-update-icon-cache', '-q', '-t' ,'-f', hicolor])
+    if subprocess.call(['gtk-update-icon-cache', '-q', '-t' ,'-f', hicolor]) == 0:
+        print('Done')
+    else: 
+        print('Error')
+
diff --git a/src/Application.vala b/src/Application.vala
index 2c043c8..36fd9a4 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -83,6 +83,8 @@ public class LightPadWindow : Widgets.CompositedWindow {
         } else if (suggested_size >= 56) {
             this.icon_size = 64;
         }
+        // phineas
+        //  this.icon_size = 24;
         message ("The apps icon size is: %d", this.icon_size);
 
         // Get all apps
@@ -135,7 +137,10 @@ public class LightPadWindow : Widgets.CompositedWindow {
             this.grid_y = 6;
             this.grid_x = 5;
         }
-
+        // phineas
+        //  this.grid_y = 8;
+        //  this.grid_x = 4;
+        message ("Grid size: x=%d, y=%d", this.grid_x, this.grid_y);
         // Initialize the grid
         for (int c = 0; c < this.grid_y; c++) {
             this.grid.insert_column (c);
diff --git a/src/Config.vala b/src/Config.vala
new file mode 100644
index 0000000..a862790
--- /dev/null
+++ b/src/Config.vala
@@ -0,0 +1,109 @@
+class Config {
+    // Grid
+    public int grid_y;
+    public int grid_x;
+    public int grid_cell_spacing_w;
+    public int grid_cell_spacing_h;
+    // AppItem
+    public double item_font_size;
+    public int item_icon_size;
+    public int item_box_width;
+    public int item_box_height;
+    // SearchBar
+    public int sb_width;
+    public int sb_height;
+
+    private int screen_w;
+    private int screen_h;
+    private KeyFile config_f;
+
+    public Config(int screen_width, int screen_height) {
+        screen_w = screen_width;
+        screen_w = screen_height;
+        config_f = new KeyFile();
+        // This could throw if file not found => app would crash
+        config_f.load_from_file(Resources.CONFIG_FILE, KeyFileFlags.KEEP_COMMENTS);
+        
+        const string[] group = {"Grid", "AppItem", "SearchBar"};
+        
+        grid_y              = config_f.get_integer(group[0], "Y");
+        grid_x              = config_f.get_integer(group[0], "X");
+        grid_cell_spacing_w = config_f.get_integer(group[0], "CellSpacingWidth");
+        grid_cell_spacing_h = config_f.get_integer(group[0], "CellSpacingHeight");
+
+        item_font_size      = config_f.get_double(group[1], "FontSize");
+        item_icon_size      = config_f.get_integer(group[1], "IconSize");
+        item_box_width      = config_f.get_integer(group[1], "BoxWidth");
+        item_box_height     = config_f.get_integer(group[1], "BoxHeight");
+
+        sb_width            = config_f.get_integer(group[2], "Width");
+        sb_height           = config_f.get_integer(group[2], "Height");
+        
+    }
+
+    private void set_config_values() {
+        if (item_icon_size != -1) {
+            set_based_on_icon_size();
+        }
+    }
+
+    private void set_based_on_icon_size() {
+        if (item_box_width == -1 && item_box_height == -1) {
+
+        }
+    }
+
+    private void set_default_icon_size() {
+        double scale_factor = (1.0/3.0);
+        double suggested_size = Math.pow (screen_w * screen_w, scale_factor);
+        suggested_size = suggested_size / 1.7;
+        if (suggested_size < 27) {
+            this.item_icon_size = 24;
+        } else if (suggested_size >= 27 && suggested_size < 40) {
+            this.item_icon_size = 32;
+        } else if ((suggested_size >= 40 && suggested_size < 56) || (screen_h == 720)) {
+            this.item_icon_size = 48;
+        } else if (suggested_size >= 56) {
+            this.item_icon_size = 64;
+        }
+    }
+
+    private void set_default_grid_size() {
+        int x, y;
+        // For Monitor 5:4 and 4:3
+        if ((screen_w / (double) screen_h) < 1.4) {
+            x = 5;
+            y = 5;
+        } else if (screen_h == 600) { // Netbook 1024x600px
+            y = 6;
+            x = 4;
+        } else if (screen_h == 720) { // HD 1280x720px
+            y = 7;
+            x = 5;
+        } else if (screen_h == 1080) { // Full HD 1920x1080px
+            y = 9;
+            x = 7;
+        } else { // Monitor 16:9
+            y = 6;
+            x = 5;
+        }
+
+        if (grid_x == -1)
+            grid_x = x;
+        if (grid_y == -1)
+            grid_y = y;
+    }
+
+    private void set_default_box_size() {
+        if (item_icon_size == -1)
+            set_default_icon_size();
+        if (item_box_width == -1)
+            item_box_width = item_icon_size * 3;
+        if (item_box_height == -1)
+            item_box_height = item_icon_size + 30;
+    }
+
+    private void set_default_font_size() {
+        item_font_size = 11.5;
+    }
+}
\ No newline at end of file
diff --git a/src/Resources.vala b/src/Resources.vala
index b0223ed..b51e01f 100644
--- a/src/Resources.vala
+++ b/src/Resources.vala
@@ -25,4 +25,5 @@
 namespace Resources {
      public const string LIGHTPAD_CONFIG_DIR = "/." + Config.PROJECT_NAME;
      public const string BLACKLIST_FILE = LIGHTPAD_CONFIG_DIR + "/blacklist";
+     public const string CONFIG_FILE = LIGHTPAD_CONFIG_DIR + "/config";
 }
diff --git a/src/Widgets/AppItem.vala b/src/Widgets/AppItem.vala
index 67725af..bdbb384 100644
--- a/src/Widgets/AppItem.vala
+++ b/src/Widgets/AppItem.vala
@@ -37,10 +37,11 @@ namespace LightPad.Frontend {
         public AppItem (int size) {
             this.icon_size = size;
 
-            // EventBox Properties
+            // EventBox Properties, a box that show up on hover
             this.set_visible_window (false);
             this.can_focus = true;
             // 30 is the padding between icon and label's height
+            // phineas
             this.set_size_request (icon_size * 3, icon_size + 30);
 
             // VBox properties
@@ -110,6 +111,8 @@ namespace LightPad.Frontend {
             Cairo.TextExtents extents;
             context.select_font_face ("Sans", Cairo.FontSlant.NORMAL, Cairo.FontWeight.NORMAL);
             context.set_font_size (11.5);
+            // phineas
+            //  context.set_font_size(32);
             LightPad.Frontend.Utilities.truncate_text (context, size, 10, this.label, out this.label, out extents);
 
             // Draw text shadow
diff --git a/src/config b/src/config
new file mode 100644
index 0000000..e9a9fee
--- /dev/null
+++ b/src/config
@@ -0,0 +1,15 @@
+[Grid]
+Y=-1
+X=-1
+CellSpacingWidth=-1
+CellSpacingHeight=-1
+
+[AppItem]
+FontSize=-1.0
+IconSize=-1
+BoxWidth=-1
+BoxHeight=-1
+
+[SearchBar]
+Width=-1
+Height=-1
\ No newline at end of file

From 5accddd016d245e010e7c21644bd6e1e5a5b7f58 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ph=C3=BAc=20La?= <phucla@outlook.com>
Date: Wed, 19 May 2021 11:45:17 +0700
Subject: [PATCH 06/17] rewrite config

---
 meson.build          |   1 +
 src/Application.vala |   5 ++
 src/Config.vala      | 146 ++++++++++++++++++++++++-------------------
 3 files changed, 89 insertions(+), 63 deletions(-)

diff --git a/meson.build b/meson.build
index 32b5677..2f3948e 100644
--- a/meson.build
+++ b/meson.build
@@ -47,6 +47,7 @@ executable(
     'src/Utilities.vala',
     'src/Color.vala',
     'src/Resources.vala',
+    'src/Config.vala',
     'src/Widgets/CompositedWindow.vala',
     'src/Widgets/Searchbar.vala',
     'src/Widgets/Indicators.vala',
diff --git a/src/Application.vala b/src/Application.vala
index 36fd9a4..fbaaabe 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -59,6 +59,11 @@ public class LightPadWindow : Widgets.CompositedWindow {
         monitor_dimensions.width = pixel_geo.width / monitor.get_scale_factor ();
         monitor_dimensions.height = pixel_geo.height / monitor.get_scale_factor ();
 
+        // phineas
+        FileConfig config = new FileConfig(monitor_dimensions.width, monitor_dimensions.height, user_home + Resources.CONFIG_FILE);
+
+        message("Icon size config: %d", config.item_icon_size);
+
         // Window properties
         this.set_title ("LightPad");
         /* Skip that a workspace switcher and taskbars displays a
diff --git a/src/Config.vala b/src/Config.vala
index a862790..22a021f 100644
--- a/src/Config.vala
+++ b/src/Config.vala
@@ -1,4 +1,4 @@
-class Config {
+class BaseConfig {
     // Grid
     public int grid_y;
     public int grid_x;
@@ -13,47 +13,21 @@ class Config {
     public int sb_width;
     public int sb_height;
 
-    private int screen_w;
-    private int screen_h;
-    private KeyFile config_f;
+    public int screen_w;
+    public int screen_h;
 
-    public Config(int screen_width, int screen_height) {
+    public BaseConfig(int screen_width, int screen_height, bool init_default = true) {
         screen_w = screen_width;
-        screen_w = screen_height;
-        config_f = new KeyFile();
-        // This could throw if file not found => app would crash
-        config_f.load_from_file(Resources.CONFIG_FILE, KeyFileFlags.KEEP_COMMENTS);
-        
-        const string[] group = {"Grid", "AppItem", "SearchBar"};
-        
-        grid_y              = config_f.get_integer(group[0], "Y");
-        grid_x              = config_f.get_integer(group[0], "X");
-        grid_cell_spacing_w = config_f.get_integer(group[0], "CellSpacingWidth");
-        grid_cell_spacing_h = config_f.get_integer(group[0], "CellSpacingHeight");
-
-        item_font_size      = config_f.get_double(group[1], "FontSize");
-        item_icon_size      = config_f.get_integer(group[1], "IconSize");
-        item_box_width      = config_f.get_integer(group[1], "BoxWidth");
-        item_box_height     = config_f.get_integer(group[1], "BoxHeight");
-
-        sb_width            = config_f.get_integer(group[2], "Width");
-        sb_height           = config_f.get_integer(group[2], "Height");
-        
-    }
-
-    private void set_config_values() {
-        if (item_icon_size != -1) {
-            set_based_on_icon_size();
-        }
-    }
-
-    private void set_based_on_icon_size() {
-        if (item_box_width == -1 && item_box_height == -1) {
-
+        screen_h = screen_height;
+        if (init_default) {
+            default_icon_size();
+            default_box_size();
+            default_font_size();
+            default_grid_size();
         }
     }
 
-    private void set_default_icon_size() {
+    private void default_icon_size() {
         double scale_factor = (1.0/3.0);
         double suggested_size = Math.pow (screen_w * screen_w, scale_factor);
         suggested_size = suggested_size / 1.7;
@@ -68,42 +42,88 @@ class Config {
         }
     }
 
-    private void set_default_grid_size() {
-        int x, y;
+    private void default_grid_size() {
         // For Monitor 5:4 and 4:3
         if ((screen_w / (double) screen_h) < 1.4) {
-            x = 5;
-            y = 5;
+            grid_x = 5;
+            grid_y = 5;
         } else if (screen_h == 600) { // Netbook 1024x600px
-            y = 6;
-            x = 4;
+            grid_y = 6;
+            grid_x = 4;
         } else if (screen_h == 720) { // HD 1280x720px
-            y = 7;
-            x = 5;
+            grid_y = 7;
+            grid_x = 5;
         } else if (screen_h == 1080) { // Full HD 1920x1080px
-            y = 9;
-            x = 7;
+            grid_y = 9;
+            grid_x = 7;
         } else { // Monitor 16:9
-            y = 6;
-            x = 5;
+            grid_y = 6;
+            grid_x = 5;
         }
-
-        if (grid_x == -1)
-            grid_x = x;
-        if (grid_y == -1)
-            grid_y = y;
     }
 
-    private void set_default_box_size() {
-        if (item_icon_size == -1)
-            set_default_icon_size();
-        if (item_box_width == -1)
-            item_box_width = item_icon_size * 3;
-        if (item_box_height == -1)
-            item_box_height = item_icon_size + 30;
+    private void default_box_size() {
+        item_box_width = item_icon_size * 3;
+        item_box_height = item_icon_size + 30;
     }
 
-    private void set_default_font_size() {
+    private void default_font_size() {
         item_font_size = 11.5;
     }
-}
\ No newline at end of file
+}
+
+void merge_int(int* ptr, int val) {
+    if (val > -1)
+        *ptr = val;
+}
+
+void merge_double(double* ptr, double val) {
+    if (val > -1.0)
+        *ptr = val;
+}
+
+class FileConfig : BaseConfig {
+    private KeyFile config_f;
+
+    public FileConfig(int screen_width, int screen_height, string file) {
+        base(screen_width, screen_height);
+
+        config_f = new KeyFile();
+        try {
+            config_f.load_from_file(file, KeyFileFlags.KEEP_COMMENTS);
+        } catch {
+            message ("Config file not found. Using default values");
+            return;
+        }
+        
+        const string[] group = {"Grid", "AppItem", "SearchBar"};
+        try {
+            merge_int(&grid_y, config_f.get_integer(group[0], "Y"));
+            merge_int(&grid_x, config_f.get_integer(group[0], "X"));
+            merge_int(&grid_cell_spacing_w, config_f.get_integer(group[0], "CellSpacingWidth"));
+            merge_int(&grid_cell_spacing_h, config_f.get_integer(group[0], "CellSpacingHeight"));
+
+            merge_double(&item_font_size, config_f.get_double(group[1], "FontSize"));
+            merge_int(&item_icon_size, config_f.get_integer(group[1], "IconSize"));
+            merge_int(&item_box_width, config_f.get_integer(group[1], "BoxWidth"));
+            merge_int(&item_box_height, config_f.get_integer(group[1], "BoxHeight"));
+
+            merge_int(&sb_width, config_f.get_integer(group[2], "Width"));
+            merge_int(&sb_height, config_f.get_integer(group[2], "Height"));
+        }
+        catch {
+            message ("Key config missing");
+        }
+    }
+}
+
+        //  grid_y              = config_f.get_integer(group[0], "Y");
+        //  grid_x              = config_f.get_integer(group[0], "X");
+        //  grid_cell_spacing_w = config_f.get_integer(group[0], "CellSpacingWidth");
+        //  grid_cell_spacing_h = config_f.get_integer(group[0], "CellSpacingHeight");
+        //  item_font_size      = config_f.get_double(group[1], "FontSize");
+        //  item_icon_size      = config_f.get_integer(group[1], "IconSize");
+        //  item_box_width      = config_f.get_integer(group[1], "BoxWidth");
+        //  item_box_height     = config_f.get_integer(group[1], "BoxHeight");
+        //  sb_width            = config_f.get_integer(group[2], "Width");
+        //  sb_height           = config_f.get_integer(group[2], "Height");
\ No newline at end of file

From 1ca0e8ac0576dcc0b4d2f136b2273182c1ef5c38 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ph=C3=BAc=20La?= <phucla@outlook.com>
Date: Wed, 19 May 2021 12:01:54 +0700
Subject: [PATCH 07/17] move default icon size and grid size to config

---
 src/Application.vala | 52 ++++++++++----------------------------------
 1 file changed, 11 insertions(+), 41 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index fbaaabe..37697c5 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -59,10 +59,18 @@ public class LightPadWindow : Widgets.CompositedWindow {
         monitor_dimensions.width = pixel_geo.width / monitor.get_scale_factor ();
         monitor_dimensions.height = pixel_geo.height / monitor.get_scale_factor ();
 
-        // phineas
-        FileConfig config = new FileConfig(monitor_dimensions.width, monitor_dimensions.height, user_home + Resources.CONFIG_FILE);
+        FileConfig config = new FileConfig(
+            monitor_dimensions.width, 
+            monitor_dimensions.height, 
+            user_home + Resources.CONFIG_FILE
+        );
+        // For compatibility, maybe add FileConfig to LightPadWindow someday
+        this.icon_size = config.item_icon_size;
+        this.grid_y = config.grid_y;
+        this.grid_x = config.grid_x;
 
-        message("Icon size config: %d", config.item_icon_size);
+        message ("The apps icon size is: %d", this.icon_size);
+        message ("The grid size are: %dx%d", this.grid_y, this.grid_x);
 
         // Window properties
         this.set_title ("LightPad");
@@ -75,22 +83,6 @@ public class LightPadWindow : Widgets.CompositedWindow {
         message ("The monitor dimensions are: %dx%d", monitor_dimensions.width,  monitor_dimensions.height);
         this.set_default_size (monitor_dimensions.width,  monitor_dimensions.height);
 
-        // Set apps icon size
-        double scale_factor = (1.0/3.0);
-        double suggested_size = Math.pow (monitor_dimensions.width * monitor_dimensions.height, scale_factor);
-        suggested_size = suggested_size / 1.7;
-        if (suggested_size < 27) {
-            this.icon_size = 24;
-        } else if (suggested_size >= 27 && suggested_size < 40) {
-            this.icon_size = 32;
-        } else if ((suggested_size >= 40 && suggested_size < 56) || (monitor_dimensions.height == 720)) {
-            this.icon_size = 48;
-        } else if (suggested_size >= 56) {
-            this.icon_size = 64;
-        }
-        // phineas
-        //  this.icon_size = 24;
-        message ("The apps icon size is: %d", this.icon_size);
 
         // Get all apps
         LightPad.Backend.DesktopEntries.enumerate_apps (this.icons, this.icon_size, user_home, out this.apps);
@@ -124,28 +116,6 @@ public class LightPadWindow : Widgets.CompositedWindow {
         this.grid.set_column_spacing (0);
         this.grid.set_halign (Gtk.Align.CENTER);
 
-        // Make icon grid and populate
-        // For Monitor 5:4 and 4:3
-        if ((monitor_dimensions.width / (double) monitor_dimensions.height) < 1.4) {
-            this.grid_x = 5;
-            this.grid_y = 5;
-        } else if (monitor_dimensions.height == 600) { // Netbook 1024x600px
-            this.grid_y = 6;
-            this.grid_x = 4;
-        } else if (monitor_dimensions.height == 720) { // HD 1280x720px
-            this.grid_y = 7;
-            this.grid_x = 5;
-        } else if (monitor_dimensions.height == 1080) { // Full HD 1920x1080px
-            this.grid_y = 9;
-            this.grid_x = 7;
-        } else { // Monitor 16:9
-            this.grid_y = 6;
-            this.grid_x = 5;
-        }
-        // phineas
-        //  this.grid_y = 8;
-        //  this.grid_x = 4;
-        message ("Grid size: x=%d, y=%d", this.grid_x, this.grid_y);
         // Initialize the grid
         for (int c = 0; c < this.grid_y; c++) {
             this.grid.insert_column (c);

From 5804375b097b979143787f10ef71e6dd517925b8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ph=C3=BAc=20La?= <phucla@outlook.com>
Date: Wed, 19 May 2021 12:24:57 +0700
Subject: [PATCH 08/17] grid spacing now adjustable

---
 src/Application.vala |  8 ++++----
 src/Config.vala      | 31 ++++++++++++++-----------------
 src/config           |  4 ++--
 3 files changed, 20 insertions(+), 23 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index 37697c5..68aac86 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -69,6 +69,7 @@ public class LightPadWindow : Widgets.CompositedWindow {
         this.grid_y = config.grid_y;
         this.grid_x = config.grid_x;
 
+        message ("The monitor dimensions are: %dx%d", monitor_dimensions.width,  monitor_dimensions.height);
         message ("The apps icon size is: %d", this.icon_size);
         message ("The grid size are: %dx%d", this.grid_y, this.grid_x);
 
@@ -80,7 +81,6 @@ public class LightPadWindow : Widgets.CompositedWindow {
         this.set_skip_taskbar_hint (true);
         this.set_type_hint (Gdk.WindowTypeHint.NORMAL);
         this.fullscreen ();
-        message ("The monitor dimensions are: %dx%d", monitor_dimensions.width,  monitor_dimensions.height);
         this.set_default_size (monitor_dimensions.width,  monitor_dimensions.height);
 
 
@@ -108,12 +108,12 @@ public class LightPadWindow : Widgets.CompositedWindow {
         int screen_half = (monitor_dimensions.width / 2) - 120;
         bottom.pack_start (this.searchbar, false, true, screen_half);
 
-        // Upstairs
+        // Upstairs (padding is the space between search bar and the grid)
         container.pack_start (bottom, false, true, 32);
 
         this.grid = new Gtk.Grid();
-        this.grid.set_row_spacing (30);
-        this.grid.set_column_spacing (0);
+        this.grid.set_row_spacing (config.grid_row_spacing);
+        this.grid.set_column_spacing (config.grid_col_spacing);
         this.grid.set_halign (Gtk.Align.CENTER);
 
         // Initialize the grid
diff --git a/src/Config.vala b/src/Config.vala
index 22a021f..164b750 100644
--- a/src/Config.vala
+++ b/src/Config.vala
@@ -2,8 +2,8 @@ class BaseConfig {
     // Grid
     public int grid_y;
     public int grid_x;
-    public int grid_cell_spacing_w;
-    public int grid_cell_spacing_h;
+    public int grid_row_spacing;
+    public int grid_col_spacing;
     // AppItem
     public double item_font_size;
     public int item_icon_size;
@@ -21,9 +21,15 @@ class BaseConfig {
         screen_h = screen_height;
         if (init_default) {
             default_icon_size();
-            default_box_size();
-            default_font_size();
             default_grid_size();
+
+            item_font_size = 11.5;
+
+            item_box_width = item_icon_size * 3;
+            item_box_height = item_icon_size + 30;
+
+            grid_row_spacing = 30;
+            grid_col_spacing = 0;
         }
     }
 
@@ -61,15 +67,6 @@ class BaseConfig {
             grid_x = 5;
         }
     }
-
-    private void default_box_size() {
-        item_box_width = item_icon_size * 3;
-        item_box_height = item_icon_size + 30;
-    }
-
-    private void default_font_size() {
-        item_font_size = 11.5;
-    }
 }
 
 void merge_int(int* ptr, int val) {
@@ -100,8 +97,8 @@ class FileConfig : BaseConfig {
         try {
             merge_int(&grid_y, config_f.get_integer(group[0], "Y"));
             merge_int(&grid_x, config_f.get_integer(group[0], "X"));
-            merge_int(&grid_cell_spacing_w, config_f.get_integer(group[0], "CellSpacingWidth"));
-            merge_int(&grid_cell_spacing_h, config_f.get_integer(group[0], "CellSpacingHeight"));
+            merge_int(&grid_row_spacing, config_f.get_integer(group[0], "RowSpacing"));
+            merge_int(&grid_col_spacing, config_f.get_integer(group[0], "ColumnSpacing"));
 
             merge_double(&item_font_size, config_f.get_double(group[1], "FontSize"));
             merge_int(&item_icon_size, config_f.get_integer(group[1], "IconSize"));
@@ -119,8 +116,8 @@ class FileConfig : BaseConfig {
 
         //  grid_y              = config_f.get_integer(group[0], "Y");
         //  grid_x              = config_f.get_integer(group[0], "X");
-        //  grid_cell_spacing_w = config_f.get_integer(group[0], "CellSpacingWidth");
-        //  grid_cell_spacing_h = config_f.get_integer(group[0], "CellSpacingHeight");
+        //  grid_row_spacing = config_f.get_integer(group[0], "RowSpacing");
+        //  grid_col_spacing = config_f.get_integer(group[0], "ColumnSpacing");
         //  item_font_size      = config_f.get_double(group[1], "FontSize");
         //  item_icon_size      = config_f.get_integer(group[1], "IconSize");
         //  item_box_width      = config_f.get_integer(group[1], "BoxWidth");
diff --git a/src/config b/src/config
index e9a9fee..ab32210 100644
--- a/src/config
+++ b/src/config
@@ -1,8 +1,8 @@
 [Grid]
 Y=-1
 X=-1
-CellSpacingWidth=-1
-CellSpacingHeight=-1
+RowSpacing=-1
+ColumnSpacing=-1
 
 [AppItem]
 FontSize=-1.0

From 672ebe6dc137862049cfd33e2e699005c9ec3bbe Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ph=C3=BAc=20La?= <phucla@outlook.com>
Date: Wed, 19 May 2021 12:55:28 +0700
Subject: [PATCH 09/17] AppItem now adjustable

---
 src/Application.vala     | 12 +++++++++++-
 src/Widgets/AppItem.vala | 13 ++++++-------
 2 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index 68aac86..de9ce27 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -27,7 +27,11 @@ public class LightPadWindow : Widgets.CompositedWindow {
     public Gee.ArrayList<Gee.HashMap<string, string>> filtered = new Gee.ArrayList<Gee.HashMap<string, string>> ();
     public LightPad.Frontend.Indicators pages;
 
+    public double font_size;
     public int icon_size;
+    public int item_box_width;
+    public int item_box_height;
+
     public int total_pages;
     public int scroll_times = 0;
     public int SCROLL_SENSITIVITY = 12;
@@ -66,6 +70,9 @@ public class LightPadWindow : Widgets.CompositedWindow {
         );
         // For compatibility, maybe add FileConfig to LightPadWindow someday
         this.icon_size = config.item_icon_size;
+        this.font_size = config.item_font_size;
+        this.item_box_width = config.item_box_width;
+        this.item_box_height = config.item_box_height;
         this.grid_y = config.grid_y;
         this.grid_x = config.grid_x;
 
@@ -198,7 +205,10 @@ public class LightPadWindow : Widgets.CompositedWindow {
     private void populate_grid () {
         for (int r = 0; r < this.grid_x; r++) {
             for (int c = 0; c < this.grid_y; c++) {
-                var item = new LightPad.Frontend.AppItem (this.icon_size);
+                var item = new LightPad.Frontend.AppItem (
+                    this.icon_size, this.font_size, 
+                    this.item_box_width, this.item_box_height
+                );
                 this.children.append (item);
 
                 item.button_press_event.connect ( () => { item.grab_focus (); return true; } );
diff --git a/src/Widgets/AppItem.vala b/src/Widgets/AppItem.vala
index bdbb384..551da8f 100644
--- a/src/Widgets/AppItem.vala
+++ b/src/Widgets/AppItem.vala
@@ -27,6 +27,7 @@ namespace LightPad.Frontend {
         private LightPad.Frontend.Color prominent;
         private string label;
         private Gtk.Box wrapper;
+        private double font_size;
         private int icon_size;
 
         const int FPS = 24;
@@ -34,15 +35,15 @@ namespace LightPad.Frontend {
         const int RUN_LENGTH = (int)(DURATION/FPS); // Total number of frames
         private int current_frame = 1; // Run length, in frames
 
-        public AppItem (int size) {
+        public AppItem (int size, double font_size, int box_width, int box_height) {
             this.icon_size = size;
+            this.font_size = font_size;
 
             // EventBox Properties, a box that show up on hover
             this.set_visible_window (false);
             this.can_focus = true;
-            // 30 is the padding between icon and label's height
-            // phineas
-            this.set_size_request (icon_size * 3, icon_size + 30);
+            // Height is also the padding between icon and label's height
+            this.set_size_request (box_width, box_height);
 
             // VBox properties
             this.wrapper = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
@@ -110,9 +111,7 @@ namespace LightPad.Frontend {
             // Truncate text
             Cairo.TextExtents extents;
             context.select_font_face ("Sans", Cairo.FontSlant.NORMAL, Cairo.FontWeight.NORMAL);
-            context.set_font_size (11.5);
-            // phineas
-            //  context.set_font_size(32);
+            context.set_font_size (this.font_size);
             LightPad.Frontend.Utilities.truncate_text (context, size, 10, this.label, out this.label, out extents);
 
             // Draw text shadow

From 68596ac2324a62f71f10311dda68ecec7c71365c Mon Sep 17 00:00:00 2001
From: RPI News <44128563+mobilegmYT@users.noreply.github.com>
Date: Thu, 22 Jul 2021 19:58:20 -0600
Subject: [PATCH 10/17] Change twister os url

---
 README.md | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/README.md b/README.md
index 44c5f42..d6c9cda 100644
--- a/README.md
+++ b/README.md
@@ -2,7 +2,7 @@
 
 LightPad is a lightweight, simple and powerful application launcher. It is also Wayland compatible.
 
-It is being developed for [Twister OS](https://raspbian-x.com/) and in collaboration with Ubuntu Budgie (and its [BudgieLightpad Applet](https://github.com/ubuntubudgie/budgie-lightpad-applet) for the system), a big thanks to [@fossfreedom](https://github.com/fossfreedom)
+It is being developed for [Twister OS](https://twisteros.com/) and in collaboration with Ubuntu Budgie (and its [BudgieLightpad Applet](https://github.com/ubuntubudgie/budgie-lightpad-applet) for the system), a big thanks to [@fossfreedom](https://github.com/fossfreedom)
 
 This project was originally forked from Slingshot from the elementary team:
 

From 8014c3ef15abbaf2b52f7e6c48bdc9cff2b9334c Mon Sep 17 00:00:00 2001
From: Mariz <marizmelo@users.noreply.github.com>
Date: Mon, 23 Aug 2021 11:01:46 -0700
Subject: [PATCH 11/17] Update README.md

Adding instruction to clone source code before starting building process
---
 README.md | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/README.md b/README.md
index d6c9cda..603b123 100644
--- a/README.md
+++ b/README.md
@@ -30,20 +30,25 @@ $ sudo apt-get install com.github.libredeb.lightpad
    ```
       $ sudo dnf install meson ninja-build libgee-devel gnome-menus-devel cdbs vala libvala-devel glib-devel libwnck-devel gtk3-devel xterm python3 python3-wheel python3-setuptools gnome-menus
    ```
-   2. Create a build folder:
+   2. Clone this repository into your machine
+   ```
+     $ git clone https://github.com/libredeb/lightpad.git
+     $ cd lightpad/
+   ```
+   3. Create a build folder:
    ```
       $ meson build --prefix=/usr
    ```
-   3. Compile LightPad:
+   4. Compile LightPad:
    ```
       $ cd build
       $ ninja
    ```
-   4. Install LightPad in the system:
+   5. Install LightPad in the system:
    ```
       $ sudo ninja install
    ```
-   5. (OPTIONAL) Uninstall LightPad:
+   6. (OPTIONAL) Uninstall LightPad:
    ```
       $ sudo ninja uninstall
    ```

From c144d167cd9f8f6172bc625852aafe75c41404eb Mon Sep 17 00:00:00 2001
From: Juan Lozano <libredeb@gmail.com>
Date: Sun, 29 May 2022 01:17:06 -0300
Subject: [PATCH 12/17] Fixed issue #28, cant open dbus activatable apps

---
 src/Application.vala | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index de9ce27..826d594 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -233,9 +233,15 @@ public class LightPadWindow : Widgets.CompositedWindow {
                         if (this.filtered.get (app_index)["terminal"] == "true") {
                             GLib.AppInfo.create_from_commandline (this.filtered.get (app_index)["command"], null, GLib.AppInfoCreateFlags.NEEDS_TERMINAL).launch (null, null);
                         } else {
-                            new GLib.DesktopAppInfo.from_filename (this.filtered.get(app_index)["desktop_file"]).launch (null, null);
+                            var context = new AppLaunchContext ();
+                            new GLib.DesktopAppInfo.from_filename (this.filtered.get(app_index)["desktop_file"]).launch (null, context);
                         }
-                        this.destroy ();
+                        this.hide();
+                        GLib.Timeout.add_seconds (1, () => {
+                            // allow some time before quitting to allow dbusactivatable apps to be launched
+                            this.destroy ();
+                            return GLib.Source.REMOVE;
+                        });
                     } catch (GLib.Error e) {
                         warning ("Error! Load application: " + e.message);
                     }

From 0bb07c6a937995e4c9ea060b3723557588461753 Mon Sep 17 00:00:00 2001
From: Juan Lozano <libredeb@gmail.com>
Date: Sun, 29 May 2022 02:45:57 -0300
Subject: [PATCH 13/17] Fixed isse #26, wont open in primary monitor

---
 src/Application.vala | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/Application.vala b/src/Application.vala
index 826d594..1b1785b 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -87,7 +87,17 @@ public class LightPadWindow : Widgets.CompositedWindow {
         this.set_skip_pager_hint (true);
         this.set_skip_taskbar_hint (true);
         this.set_type_hint (Gdk.WindowTypeHint.NORMAL);
-        this.fullscreen ();
+        //this.fullscreen (); <-- old method used
+        var display = Gdk.Display.get_default();
+        message ("Amount of Monitors: %d", display.get_n_monitors ());
+        int primary_monitor_number = 0;
+        for (int i = 0; i < display.get_n_monitors (); i++) {
+            if (get_display ().get_monitor (i).is_primary ()) {
+                primary_monitor_number = i;
+            }
+        }
+        message ("Your primary monitor is: %d", primary_monitor_number);
+        this.fullscreen_on_monitor (monitor.get_display ().get_default_screen (), primary_monitor_number);
         this.set_default_size (monitor_dimensions.width,  monitor_dimensions.height);
 
 

From b95062dbaf3d3889eb99a5a74dd07e0bc650761c Mon Sep 17 00:00:00 2001
From: Juan Lozano <libredeb@gmail.com>
Date: Sun, 29 May 2022 02:54:12 -0300
Subject: [PATCH 14/17] Added notes in README.md

---
 README.md            | 5 +++++
 src/Application.vala | 1 -
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/README.md b/README.md
index 603b123..3f47bb4 100644
--- a/README.md
+++ b/README.md
@@ -97,6 +97,11 @@ These lines appear in the **.desktop** files located in `/usr/share/applications
 
 
 ## Changelog
+**Version 0.0.9**
+* Fixed [issue #26](https://github.com/libredeb/lightpad/issues/26), opens in wrong monitor
+* Fixed [issue #28](https://github.com/libredeb/lightpad/issues/28), can't run gnome apps
+* Release in progress... 
+
 **Version 0.0.8**
 * Templates added to make packages for Arch Linux (PKG) and Fedora (RPM)
 * Config files are introduced for project constants, replacing the hardcoded paths
diff --git a/src/Application.vala b/src/Application.vala
index 1b1785b..732b2d1 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -96,7 +96,6 @@ public class LightPadWindow : Widgets.CompositedWindow {
                 primary_monitor_number = i;
             }
         }
-        message ("Your primary monitor is: %d", primary_monitor_number);
         this.fullscreen_on_monitor (monitor.get_display ().get_default_screen (), primary_monitor_number);
         this.set_default_size (monitor_dimensions.width,  monitor_dimensions.height);
 

From a896d48ba7c5f28a2f0d5536e69a05300ee2b349 Mon Sep 17 00:00:00 2001
From: DEB Libre <libredeb@gmail.com>
Date: Mon, 30 May 2022 22:23:11 -0300
Subject: [PATCH 15/17] Added Arch Linux packages

---
 README.md | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/README.md b/README.md
index 3f47bb4..d4dfea7 100644
--- a/README.md
+++ b/README.md
@@ -30,6 +30,10 @@ $ sudo apt-get install com.github.libredeb.lightpad
    ```
       $ sudo dnf install meson ninja-build libgee-devel gnome-menus-devel cdbs vala libvala-devel glib-devel libwnck-devel gtk3-devel xterm python3 python3-wheel python3-setuptools gnome-menus
    ```
+   * For Arch Linux:
+   ```
+      $ sudo pacman -Sy meson ninja libgee gnome-menus vala glib2 gdk-pixbuf2 libwnck3 gtk3 xterm python python-wheel python-setuptools
+   ```
    2. Clone this repository into your machine
    ```
      $ git clone https://github.com/libredeb/lightpad.git

From 898ebda025392b46bb35d174a72dc2667acca570 Mon Sep 17 00:00:00 2001
From: Juan Lozano <libredeb@gmail.com>
Date: Tue, 31 May 2022 01:07:55 -0300
Subject: [PATCH 16/17] Fixed issue #23, cant leave lightpad clicking on empty
 area

---
 src/Application.vala | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/src/Application.vala b/src/Application.vala
index 732b2d1..b0aa56f 100644
--- a/src/Application.vala
+++ b/src/Application.vala
@@ -190,8 +190,37 @@ public class LightPadWindow : Widgets.CompositedWindow {
         }
 
         this.draw.connect (this.draw_background);
-        //this.focus_out_event.connect ( () => { this.destroy(); return true; } );
-
+        // close Lightpad when the window loses focus
+        this.focus_out_event.connect ( () => {
+            this.hide();
+            GLib.Timeout.add_seconds (1, () => {
+                this.destroy ();
+                return GLib.Source.REMOVE;
+            });
+            return true;
+        } );
+        // close Lightpad when we clic on empty area
+        this.button_release_event.connect ( () => {
+            // for some reason, the searchbar widget is part of the main_window (this)
+            // widget. So we can not check if the searchbar is focused or not. So this
+            // is a workaround based on the searchbar dimmensions and clicked position
+            // based on this widget.
+            int x, y;
+            int w, h;
+            this.searchbar.get_pointer(out x, out y);
+            this.searchbar.get_size_request(out w, out h);
+            if (( (x <= w) && (x >= 0) ) && ( (y <= h) && (y >= 0) )) {
+                return false;
+            } else {
+                this.hide();
+                GLib.Timeout.add_seconds (1, () => {
+                    this.destroy ();
+                    return GLib.Source.REMOVE;
+                });
+                return true;
+            }
+        } );
+        
     }
 
     private void search() {

From f31e0e06dabd8d2d7562706d5ff333723180685d Mon Sep 17 00:00:00 2001
From: DEB Libre <libredeb@gmail.com>
Date: Tue, 31 May 2022 01:12:01 -0300
Subject: [PATCH 17/17] Update README.md with last bug fixing

---
 README.md | 1 +
 1 file changed, 1 insertion(+)

diff --git a/README.md b/README.md
index d4dfea7..e5ec6b8 100644
--- a/README.md
+++ b/README.md
@@ -104,6 +104,7 @@ These lines appear in the **.desktop** files located in `/usr/share/applications
 **Version 0.0.9**
 * Fixed [issue #26](https://github.com/libredeb/lightpad/issues/26), opens in wrong monitor
 * Fixed [issue #28](https://github.com/libredeb/lightpad/issues/28), can't run gnome apps
+* Fixed [issue #23](https://github.com/libredeb/lightpad/issues/23), can't exit clicking on an empty area
 * Release in progress... 
 
 **Version 0.0.8**
