--- a/meson.build	2025-05-05 21:23:20.000000000 +0000
+++ b/meson.build	2025-05-07 02:57:31.336061450 +0000
@@ -30,7 +30,15 @@
 gtk_dep = dependency('gtk+-3.0')
 gobject_dep = dependency('gobject-2.0')
 gio_unix_dep = dependency('gio-unix-2.0')
-budgie_dep = dependency('budgie-2.0')
+
+for_wayland = get_option('for-wayland')
+
+if for_wayland == false
+        budgie_dep = dependency('budgie-1.0')
+else
+	budgie_dep = dependency('budgie-2.0')
+endif
+
 peas_dep = dependency('libpeas-gtk-1.0')
 
 subdir('src')
