From 5701fa273e2c0d00058a175fe97025655b0cc698 Mon Sep 17 00:00:00 2001
From: Joshua Strobl <me@joshuastrobl.com>
Date: Sun, 24 Mar 2024 11:55:05 +0200
Subject: [PATCH] Add am_cflags to ignore poorly generated C code from vala.
 Resolves #194.

---
 meson.build | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/meson.build b/meson.build
index 529d11108..15066e978 100644
--- a/meson.build
+++ b/meson.build
@@ -12,6 +12,13 @@ project(
     ],
 )
 
+# Vala generates bad C code and missing these on gcc 14 will cause FTBFS
+am_cflags = [
+    '-Wno-incompatible-pointer-types',
+    '-Wno-implicit-function-declaration',
+]
+add_global_arguments(am_cflags, language: 'c')
+
 budgie_screensaver  = find_program('budgie-screensaver', required: false)
 gnome_screensaver = find_program('gnome-screensaver', required: false)
 with_gnome_screensaver = get_option('with-gnome-screensaver')
