diff --git a/src/raven/meson.build b/src/raven/meson.build
index 2830e39a..9394b895 100644
--- a/src/raven/meson.build
+++ b/src/raven/meson.build
@@ -6,7 +6,6 @@ gresource = join_paths(meson.current_source_dir(), 'ui', 'budgie-raven.gresource
 libraven_resources = gnome.compile_resources(
     'budgie-raven-resources',
     gresource,
-    'budgie-daemon.gresource.xml',
     source_dir: join_paths(meson.current_source_dir(), 'ui'),
     c_name: 'budgie_raven',
 )
